File: answers/lintcode/House Robber.cpp
class Solution {
public:
    /**
     * @param A: An array of non-negative integers.
     * return: The maximum amount of money you can rob tonight
     */
    long long houseRobber(vector<int> A) {
        // write your code here
		vector<long long> dp(A.size(), 0);
		for (size_t i = 0; i < A.size(); ++ i) {
			dp[i] = max(dp[i], static_cast<long long> (A[i]));
			if (i >= 1) {
				dp[i] = max(dp[i], dp[i - 1]);
			}
			if (i >= 2) {
				dp[i] = max(dp[i], dp[i - 2] + A[i]);
			}
		}
		return dp[dp.size() - 1];
    }
};

File: answers/leetcode/House Robber/House Robber.java
class Solution {
    public int rob(int[] nums) {
        int[] maxRob = new int[nums.length];
        if (nums.length > 0) {
            maxRob[0] = nums[0];
        }
        if (nums.length > 1) {
            maxRob[1] = Math.max(nums[0], nums[1]);
        }
        for (int i = 2; i < nums.length; i++) {
            maxRob[i] = Math.max(maxRob[i - 1], maxRob[i - 2] + nums[i]);
        }
        return maxRob.length > 0 ? maxRob[maxRob.length - 1] : 0;
    }
}